src/Interpreters/Aggregator.cpp

```diff
@@ -27,12 +28,150 @@
 #include <Interpreters/JIT/CompiledExpressionCache.h>
 #include <Core/ProtocolDefines.h>
 
+#include <Parsers/ASTSelectQuery.h>
 
 namespace ProfileEvents
 {
-    extern const Event ExternalAggregationWritePart;
-    extern const Event ExternalAggregationCompressedBytes;
-    extern const Event ExternalAggregationUncompressedBytes;
+extern const Event ExternalAggregationWritePart;
+extern const Event ExternalAggregationCompressedBytes;
+extern const Event ExternalAggregationUncompressedBytes;
+extern const Event HashTablesPreallocatedElements;
+extern const Event HashTablesInitedAsTwoLevel;
+}
+
+namespace
+{
+/** Collects observed HashMap-s sizes to avoid redundant intermediate resizes.
+  */
+class HashTablesStatistics
+{
+public:
+    using Cache = DB::LRUCache<UInt64, size_t>;
+    using CachePtr = std::shared_ptr<Cache>;
+    using Params = DB::Aggregator::Params::StatsCollectingParams;
+
+    std::optional<size_t> getSizeHint(const Params & params)
+    {
+        std::lock_guard lock(mutex);
+        const auto cache = getHashTableStatsCache(params, lock);
+        if (const auto hint = cache->get(params.key))
+        {
+            LOG_DEBUG(&Poco::Logger::get("Aggregator"), "An entry for key={} found in cache: size_hint={}", params.key, *hint);
+            return *hint;
+        }
+        return std::nullopt;
+    }
+
+    void update(size_t observed_size, const Params & params)
+    {
+        std::lock_guard lock(mutex);
+        const auto cache = getHashTableStatsCache(params, lock);
+        const auto hint = cache->get(params.key);
+        // We'll maintain the maximum among all the observed values until the next prediction turns out to be too wrong.
+        if (!hint || observed_size > *hint || observed_size < *hint / 2)
```

Comment:
It's really minor thing, but I suppose it's easier to read it like that, because we immediately see range when condition is not apllied `[hint/2, hint]`

```suggestion
        if (!hint || observed_size < *hint / 2 || *hint < observed_size)
```

Reply:
done